inference-engine/thirdparty/clDNN/kernel_selector/core/cl_kernels/non_max_suppression_gpu_ref.cl

```diff
@@ -0,0 +1,571 @@
+    if (workItemId == 0) {
+        __range[localClassId][0] = 0;
+        __range[localClassId][1] = sortedBoxNum - 1;
+    } else {
+        __range[localClassId][workItemId * 2] = 0;
+        __range[localClassId][workItemId * 2 + 1] = 0;
+    }
+    barrier(CLK_LOCAL_MEM_FENCE);
+
+    int range_step = 2;
+    const int first_id = workItemId * 2;
+    for (int i = 0; i < PARTITION_STEP; ++i, range_step *= 2) {
+        if (workItemId <= i) {
+            const int begin_id = __range[localClassId][first_id];
+            const int end_id = __range[localClassId][first_id + 1];
+            const int second_id = first_id + range_step;
+
+            if (begin_id < end_id) {
+                const int pivot = FUNC_CALL(partition)(sortedBoxList, begin_id, end_id);
+                __range[localClassId][first_id     ] = begin_id;
+                __range[localClassId][first_id + 1 ] = max(pivot - 1, begin_id);
+                __range[localClassId][second_id    ] = min(pivot + 1, end_id);
+                __range[localClassId][second_id + 1] = end_id;
+            }
+        }
+
+        barrier(CLK_GLOBAL_MEM_FENCE | CLK_LOCAL_MEM_FENCE);
+    }
+
+    const int begin_id = __range[localClassId][first_id];
+    const int end_id = __range[localClassId][first_id + 1];
+    if (begin_id < end_id) {
+        FUNC_CALL(quickSortIterative)(sortedBoxList, begin_id, end_id);
+    }
+}
+#endif /* IS_STAGE_1 */
+
+#ifdef IS_STAGE_2
+KERNEL (non_max_suppression_ref_stage_2)(
+    const __global INPUT0_TYPE *boxes
+    , __global uchar *buffer0
+    , __global uchar *buffer1
+    , __global int *buffer3
+    #ifdef NUM_SELECT_PER_CLASS_TYPE
+    , const __global NUM_SELECT_PER_CLASS_TYPE *num_select_per_class
+    #endif
+    #ifdef IOU_THRESHOLD_TYPE
+    , const __global IOU_THRESHOLD_TYPE *iou_threshold
+    #endif
+    #ifdef SCORE_THRESHOLD_TYPE
+    , const __global SCORE_THRESHOLD_TYPE *score_threshold
+    #endif
+    #ifdef SOFT_NMS_SIGMA_TYPE
+    , const __global SOFT_NMS_SIGMA_TYPE *soft_nms_sigma
+    #endif
+    )
+{
+    const short batchId = get_global_id(0);
+    const ushort classId = get_global_id(1);
+
+    float scale = 0.0f;
+    if (SOFT_NMS_SIGMA_VAL > 0.0f) {
+        scale = -0.5f / SOFT_NMS_SIGMA_VAL;
+    }
+
+    __global SBOX_INFO *sortedBoxList = (__global SBOX_INFO*)&buffer0[(batchId * NUM_CLASSES + classId) * BUFFER_STRIDE];
+    int sortedBoxNum = buffer3[batchId * NUM_CLASSES + classId];
+
+    __global BOX_INFO *selectedBoxList = (__global BOX_INFO*)&buffer1[(batchId * NUM_CLASSES + classId) * BUFFER_STRIDE];
+    int selectedBoxNum = 0;
+    while (sortedBoxNum != 0 && selectedBoxNum < NUM_SELECT_PER_CLASS_VAL) {
+        SBOX_INFO next_candidate = sortedBoxList[sortedBoxNum - 1];
+        INPUT1_TYPE original_score = next_candidate.score;
+        --sortedBoxNum;
+
+        bool should_hard_suppress = false;
+        for (int j = selectedBoxNum - 1;
```

Comment:
Do you mean parallelization by processing multiple WIs? I think that would be too small to parallelize among multiple WIs.

Reply:
It depends of selectedBoxNum. This number of boxes is tested with each new candidate

Reply:
That is interesting approach to parallelize IOU part. However, as mentioned above, the amount of work might be too small and it will cause additional overhead of synchronization. Nested parallelism is also required to implement it in a clean way. Let us consider to implement this when we have more clear performance target for NMS.